### ISSUE TITLE
A data-dependent store mask degenerates the 2D block store to one message and one `alloca` per row (2x on a grouped GEMM)

### ISSUE THREAD

**Issue by etaf** [intel/intel-xpu-backend-for-triton#7923]
## Summary

When a `tl.store` mask is not constant along the row dimension, `getBlockIOTileSize`
clamps the store tile height to 1 but the store still takes the 2D block store path.
A 256x256 bf16 tile is then covered one row per message: the IR carries 131
`__spirv_Subgroup2DBlockStoreINTEL` calls instead of 19, and because the store
lowering stages every message's payload through its own `alloca`, 131 stack slots
instead of 19. That is 2x slower than the `__spirv_PredicatedStoreINTEL` fallback
that the same kernel takes when the tile geometry happens to be rejected for an
unrelated reason.

This is not a corner case for MoE: a grouped GEMM must mask partial m-tiles at every
group boundary, and the bound comes from the offsets tensor, so it is data-dependent
by construction.

Measured on Arc B580 (Xe2, 20 Xe-cores), Triton `b5611e559`, `2048x4096x7168` bf16,
tile `256x256x32`, `num_stages=2`, `num_warps=32`, 20 workgroups:

| variant | ms | TFLOPS | `alloca` | `Subgroup2DBlockStore` | `PredicatedStore` |
|---|---|---|---|---|---|
| mask bound loaded from memory, constexpr row stride | **2.579** | **46.6** | **131** | **131** | 0 |
| same, row stride passed as an argument | 1.283 | 93.7 | 3 | 0 | 131 |
| mask bound as an `int` argument, constexpr row stride | 1.281 | 93.9 | 19 | 19 | 0 |
| no mask, constexpr row stride | 1.278 | 94.1 | 19 | 19 | 0 |

Only the first row is affected, and it is the only row whose mask has a
row-constancy of 1. Rows 2-4 are all within noise of each other, so the loss is
specifically the degenerate geometry, not masking and not the predicated path.

## Repro

The script at the end of this issue is self-contained (PyTorch only for tensor
allocation, no Inductor) and prints the table above. The four variants differ only in
constexpr switches around one `tl.store`.

Two details are load-bearing:

* **The mask bound must come from memory.** As a plain `int` argument, Triton's
  divisibility-by-16 specialization gives the comparison a row-constancy of 16, the
  tile survives, and the problem disappears (row 3). Real grouped GEMMs load the
  bound from the offsets tensor.
* **The output row stride must be a compile-time constant.** With a runtime stride
  the store is not recognized as row-major block I/O at all and falls back (row 2).
  Inductor always emits a constant stride (`xindex = idx_n + 4096*idx_m`), so the
  bad path is the one production code hits.

## Mechanism

1. `getBlockIOTileSize` (`third_party/intel/lib/TritonIntelGPUTransforms/BlockIOUtils.cpp:225-241`)
   clamps the row-dim tile extent to the mask's constancy along that dim:

   ```cpp
   unsigned maskConstancyRowDimLimit = std::numeric_limits<unsigned>::max();
   if (rowDim >= 0) {
     if (maskAxisInfo && !llvm::isPowerOf2_64(maskAxisInfo->getConstancy(rowDim)))
       return BlockIOTileSizeInfo::unknown();
     if (maskAxisInfo)
       maskConstancyRowDimLimit = maskAxisInfo->getConstancy(rowDim);
   }
   rowDimLimit = std::min(rowDimLimit, maskConstancyRowDimLimit);
   ```

   `offs_m[:, None] < tl.load(bound_ptr)` has constancy 1 along rows, so
   `tileHeight` ends up 1. (Confirmed by the guard below: rejecting
   `tileHeight <= 1` is what moves this variant off the 2D path.)

2. `validate2DBlockStoreTile` (`BlockIOUtils.cpp:593`) accepts that geometry: a
   1-row tile is expressible by the hardware, so none of the existing checks
   (transpose, address-payload restriction) reject it.

3. `TritonMatrix2DBlockStoreLowering` (`third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1000`)
   creates one `alloca` per store op, because the SPIR-V builtin takes the payload
   by pointer:

   ```cpp
   auto storeValPtr = LLVM::AllocaOp::create(
       rewriter, loc, ptr_ty(ctx), storeValType.getElementType(),
       b.i32_val(storeValType.getNumElements()));
   LLVM::StoreOp::create(rewriter, loc, op.getStoredVal(), storeValPtr);
   ```

   With `tileHeight == 1` there are 131 such ops instead of 19, so ~112 extra stack
   slots. They are not promoted away.

**The cost is static, not per-execution.** We also tried a template-side workaround
that masks only the ragged tail tile and leaves the full tiles unmasked, hoping the
predicated branch would be cold. It does not help (2.617 ms vs 2.746 ms): the stack
frame is reserved whether or not the masked branch runs, and it is the private-memory
pressure, not the store messages themselves, that costs the 2x. At `num_warps=16`
the same mechanism costs 11x (17.8 ms). `optimize_epilogue`, `code_sinking` and
`grf_mode` do not change it either.

## What we did locally, and why we are asking rather than sending a PR

We are running with a three-line guard in `validate2DBlockStoreTile`, right next to
the existing "the hardware cannot do this" checks:

```cpp
  // 2D block store does not support transpose.
  if (sizeInfo.transpose)
    return false;

+ if (sizeInfo.tileHeight <= 1)
+   return false;
```

Rationale: a 1-row tile is not *unstorable*, it is just unconditionally worse than
the fallback that the surrounding code already implements, and eligibility for the
2D path is already centralized in this function. It recovers the full 2x
(2.579 -> 1.277 ms) and leaves everything else alone -- the unmasked variant keeps
its 19 block stores in both builds.

It is a heuristic in a validity function, though, and we found one case where it
loses: with `BLOCK_K=16` on a store-heavy shape (`16384x7168x2048`, G=8) the same
kernel goes 6.951 -> ~10.3 ms, i.e. there the many small block stores were earning
their stack traffic back. So the flat threshold is wrong in at least one direction.

Questions:

1. Is the intended fix in the lowering instead -- hoist/reuse a single payload
   `alloca` across the messages of one tile (or emit one multi-row store and
   predicate rows), so that the degenerate geometry stops being expensive?
2. If the decision stays in `validate2DBlockStoreTile`, what should the condition
   be? `tileHeight == 1 && tileWidth * elemSize < some threshold` would keep the
   `BLOCK_K=16` case, but we have no visibility into which side is right in general.
3. Is it worth improving the mask-constancy input instead? For a grouped GEMM the
   mask is `offs_m < bound` with `offs_m` a contiguous `arange` -- the store is a
   contiguous *prefix* of rows, which a multi-row block store plus a row-count
   argument could express exactly, without per-row predication.

## Context

Found while bringing up an XPU-specific `grouped_mm` Triton template for PyTorch
Inductor on B580 (DeepSeekV3 MoE shapes, `A(Mtot,K) x B(G,K,N) -> C(Mtot,N)`).
Before the change, `BLOCK_K=32` was unusable and autotune always fell back to
`BLOCK_K=16`; after it, `256x256x32` is the best config on all six shapes we
benchmark and best-per-shape improves 3-12% end to end (geomean 1.08x), from
0.71-0.81 to 0.77-0.88 of the per-shape attainable roofline.

<details>
<summary>repro_intel_masked_store.py</summary>

```python
"""Minimal repro for the Intel Triton issue: a row-varying store mask degrades the
2D block store into one message (and one alloca) per row.

Self-contained, no inductor. One persistent bf16 GEMM epilogue, four variants:

  loaded_bound_const_stride    mask = offs_m < tl.load(bound), row stride constexpr
  loaded_bound_runtime_stride  same mask, stride passed as an argument
  int_arg_bound_const_stride   same mask, bound as an int argument instead
  unmasked_const_stride        no mask, row stride constexpr

The only difference that matters is which store path the backend picks, so the
report prints the LLVM IR counts next to the time. The bound has to come from
memory: as an int argument its divisibility-by-16 specialization gives the mask a
constancy of 16 along rows and the tile survives.

    source .env && python -u repro_intel_masked_store.py
"""

import re

import torch
import triton
import triton.language as tl


DEV = "xpu"
NUM_SMS = 20
M, N, K = 2048, 4096, 7168
BLOCK_M, BLOCK_N, BLOCK_K = 256, 256, 32
NUM_WARPS, NUM_STAGES = 32, 2


@triton.jit
def gemm(a_ptr, b_ptr, out_ptr, bound_ptr, m_size_int, out_stride_m,
         M: tl.constexpr, N: tl.constexpr, K: tl.constexpr,
         BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
         NUM_SMS: tl.constexpr, CONST_STRIDE: tl.constexpr, MASKED: tl.constexpr,
         BOUND_FROM_MEM: tl.constexpr):
    a_desc = tl.make_tensor_descriptor(a_ptr, shape=[M, K], strides=[K, 1], block_shape=[BLOCK_M, BLOCK_K])
    b_desc = tl.make_tensor_descriptor(b_ptr, shape=[K, N], strides=[N, 1], block_shape=[BLOCK_K, BLOCK_N])

    num_n_tiles: tl.constexpr = N // BLOCK_N
    num_tiles: tl.constexpr = (M // BLOCK_M) * num_n_tiles

    for tidx in range(tl.program_id(0), num_tiles, NUM_SMS):
        tile_m = tidx // num_n_tiles
        tile_n = tidx % num_n_tiles
        m_offset = tile_m * BLOCK_M
        n_offset = tile_n * BLOCK_N

        acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
        for k in range(K // BLOCK_K):
            a = a_desc.load([m_offset, k * BLOCK_K])
            b = b_desc.load([k * BLOCK_K, n_offset])
            acc += tl.dot(a, b)

        offs_m = m_offset + tl.arange(0, BLOCK_M)
        offs_n = n_offset + tl.arange(0, BLOCK_N)
        stride = N if CONST_STRIDE else out_stride_m
        xindex = offs_n[None, :] + stride * offs_m[:, None]
        m_size = tl.load(bound_ptr) if BOUND_FROM_MEM else m_size_int
        if MASKED:
            tl.store(out_ptr + xindex, acc, offs_m[:, None] < m_size)
        else:
            tl.store(out_ptr + xindex, acc)


VARIANTS = (
    ("loaded_bound_const_stride", True, True, True),
    ("loaded_bound_runtime_stride", False, True, True),
    ("int_arg_bound_const_stride", True, True, False),
    ("unmasked_const_stride", True, False, True),
)


def alloc(size, align, stream):
    return torch.empty(size, device=DEV, dtype=torch.int8)


def main():
    triton.set_allocator(alloc)
    a = torch.randn(M, K, device=DEV, dtype=torch.bfloat16) / 16
    b = torch.randn(K, N, device=DEV, dtype=torch.bfloat16)
    c = torch.empty(M, N, device=DEV, dtype=torch.bfloat16)
    ref = (a @ b).float()
    flops = 2 * M * N * K

    print(f"{torch.xpu.get_device_name()}  triton {triton.__version__}")
    print(f"{M}x{N}x{K} bf16, tile {BLOCK_M}x{BLOCK_N}x{BLOCK_K} s{NUM_STAGES} w{NUM_WARPS}, {NUM_SMS} workgroups")
    bound = torch.full((1,), M, device=DEV, dtype=torch.int32)
    print(f"{'variant':<28} {'ms':>8} {'TFLOPS':>8} {'alloca':>7} {'blockstore':>11} {'predstore':>10}   err")
    for name, const_stride, masked, bound_from_mem in VARIANTS:
        consts = dict(M=M, N=N, K=K, BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N, BLOCK_K=BLOCK_K,
                      NUM_SMS=NUM_SMS, CONST_STRIDE=const_stride, MASKED=masked,
                      BOUND_FROM_MEM=bound_from_mem, num_stages=NUM_STAGES, num_warps=NUM_WARPS)
        args = (a, b, c, bound, M, N)
        llir = gemm.warmup(*args, **consts, grid=(NUM_SMS,)).asm["llir"]
        counts = [len(re.findall(p, llir)) for p in
                  (r"= alloca", r"Subgroup2DBlockStoreINTEL", r"PredicatedStoreINTEL")]
        c.zero_()
        gemm[(NUM_SMS,)](*args, **consts)
        torch.xpu.synchronize()
        err = (c.float() - ref).abs().mean().item() / ref.abs().mean().item()
        t = triton.testing.do_bench(lambda: gemm[(NUM_SMS,)](*args, **consts))
        print(f"{name:<28} {t:8.3f} {flops / (t * 1e9):8.1f} {counts[0]:7} {counts[1]:11} {counts[2]:10}   {err:.1e}")


if __name__ == "__main__":
    main()
```

</details>

### Environment details

* Arc B580 (BMG, Xe2), 20 Xe-cores
* Triton `b5611e559` (intel-xpu-backend-for-triton), `triton 3.8.0`
* PyTorch built from source, oneAPI DPC++ 2026.1.0
* Ubuntu 26.04.1, kernel 7.0.0-30-generic

**Comment by whitneywhtsang** [comment]
The descripted issue is fixed in #7848. Please comment if need reopening.
